politeiawww/user/cockroachdb/cockroachdb.go

```diff
@@ -741,6 +742,110 @@ func (c *cockroachdb) RegisterPlugin(p user.Plugin) error {
 	return nil
 }
 
+// EmailHistoriesGet see user.Database for details.
+func (c *cockroachdb) EmailHistoriesGet(recipients []uuid.UUID) (map[uuid.UUID]user.EmailHistory, error) {
+	log.Tracef("EmailHistoriesGet: %v", recipients)
+
+	if c.isShutdown() {
+		return nil, user.ErrShutdown
+	}
+
+	if len(recipients) > user.EmailHistoriesPageLimit {
+		return nil, fmt.Errorf("too many recipients: %d, can't handle more than: %d at a time",
+			len(recipients), user.EmailHistoriesPageLimit)
+	}
+
+	var dbResult []EmailHistory
+	err := c.userDB.
+		Where("user_id IN (?)", recipients).
+		Find(&dbResult).
+		Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	result := make(map[uuid.UUID]user.EmailHistory, len(dbResult))
+	for _, dbEntity := range dbResult {
+		entity, err := c.convertEmailHistoryFromDB(dbEntity)
+		if err != nil {
+			return nil, fmt.Errorf("convert email history from db: %w", err)
+		}
+		result[dbEntity.UserID] = *entity
+	}
+
+	return result, nil
+}
+
+// EmailHistoriesSave see user.Database for details.
+func (c *cockroachdb) EmailHistoriesSave(histories map[uuid.UUID]user.EmailHistory) error {
+	log.Tracef("EmailHistoriesSave: %v %v", histories)
+
+	if c.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	if len(histories) > user.EmailHistoriesPageLimit {
+		return fmt.Errorf("too many recipients: %d, can't handle more than: %d at a time",
+			len(histories), user.EmailHistoriesPageLimit)
+	}
+
+	// Not optimal performance-wise to work with histories one by one, but very straightforward.
+	for userID, history := range histories {
+		historyDB, err := c.convertEmailHistoryToDB(userID, history)
+		if err != nil {
+			return fmt.Errorf("convert email history to DB: %w", err)
+		}
+
+		var update bool
+		h := EmailHistory{
+			UserID: userID,
+		}
+		err = c.userDB.Find(&h).Error
+		switch err {
+		case nil:
+			// DB entry already exists, update it.
+			update = true
+		case gorm.ErrRecordNotFound:
+			// DB entry doesn't exist, will have to create it.
+		default:
+			// All other errors
+			return fmt.Errorf("find email history: %w", err)
+		}
+
+		if update {
+			err := c.userDB.Save(&historyDB).Error
+			if err != nil {
+				return fmt.Errorf("save: %w", err)
+			}
+		} else {
+			err := c.userDB.Create(&historyDB).Error
+			if err != nil {
+				return fmt.Errorf("create: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *cockroachdb) convertEmailHistoryToDB(userID uuid.UUID, h user.EmailHistory) (EmailHistory, error) {
```

Comment:
What's the purpose of having these functions as methods on the `cockraochdb` context?

Reply:
no purpose really, just copied from `func (c *cockroachdb) convertSessionFromUser(s user.Session) (*Session, error)` definition,

removed.